### ISSUE TITLE
twisted.internet.task.Clock does not implement IReactorTime

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2783]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#2783|
|Type|defect|
|Created|2007-08-11 22:14:57Z|
|Branch|https://github.com/twisted/twisted/tree/clock-ireactortime-2783-2|

Specifically it has no implements(IReactorTime) declaration and doesn't implement getDelayedCalls.

<details><summary>Searchable metadata</summary>

```
trac-id__2783 2783
type__defect defect
reporter__dreid dreid
priority__highest highest
milestone__ 
branch__branches_clock_ireactortime_2783_2 branches/clock-ireactortime-2783-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1186870497000000 1186870497000000
changetime__1203541625000000 1203541625000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_1"></a>|@dreid removed owner|
|-|-|

Addressed in [branches/clock-ireactortime-2783](branches/clock-ireactortime-2783)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml set owner to @dreid|
|-|-|

Thanks for doing this dreid.

Your branch mostly looks good. It would have been helpful if you explained why `Clock` should implement `IReactorTime` in your ticket, but for something this small and obviously useful, I guess it's OK.

Minor quibbles:
 * Pyflakes says that `seconds` and `warnings` are imported into task.py without being used. Please remove these imports. (`import warnings` was added in your branch -- is there a reason for this?)
 * Please add a test that verifies that `getDelayedCalls()` returns an empty list/set/whatever on a newly-constructed `Clock`.
 * Please change your test to use `assertEqual(set([call, call2]), set(calls))`, rather than the `len`, `in`, `in` dance. This makes it clearer that you are testing for equality without regard for ordering.

Once these are done, please land it.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_3"></a>|@dreid removed owner|
|-|-|

Ok, i've addressed jml's review points.  But I'm going to put this back up for review because it's been a while and opinions might have changed.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @dreid|
|-|-|

The third point of jml's review is still true. Although I may not agree about using set, I'd prefer using sort().

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_5"></a>|@dreid removed owner|
|-|-|

Seems I accidentally made that change on my trunk working copy which was unbranched from clock-ireactortime-2783.  As to using sort instead of set.  There is no specification for sorting DelayedCalls so order is undefined and irrelevant.  

The version using set is also atleast three lines shorter than a version using sort.

Back for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @dreid|
|-|-|

Ahah caught you: python2.3 compatibility, set is not defined.

Otherwise, looks good, so please merge when corrected.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

Sorry I forget about one thing: can you add a test to show that a clock provide IReactorTime (and/or is adaptable to). Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid commented|
|-|-|

(In [22416]) Use Set from the sets module for compatibility, also add test that asserts Clock provides IReactorTime.

Refs [#2783](https://github.com/twisted/twisted/issues/2783)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @dreid|
|-|-|

Erm, in the meantime, twisted.compat.set was introduced. Thanks for merging forward and using it :).

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_11"></a>|@dreid commented|
|-|-|

(In [22637]) Branching to 'clock-ireactortime-2783-2'

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_12"></a>|@dreid commented|
|-|-|

(In [22638]) Merge forward using twisted.python.compat.set
Refs [#2783](https://github.com/twisted/twisted/issues/2783)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set owner to @dreid|
|-|-|

Great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_15"></a>|@dreid set status to `closed`|
|-|-|

(In [22640]) Merge clock-ireactortime-2783-2: Make task.Clock implement IReactorTime

Add an implements(IReactorTime) definition and implement getDelayedCalls.

Author: dreid
Reviewer: therve, jml
Fixes [#2783](https://github.com/twisted/twisted/issues/2783)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
